core/types/calls.cc

```diff
@@ -2451,6 +2451,140 @@ class Magic_selfNew : public IntrinsicMethod {
     }
 } Magic_selfNew;
 
+class Magic_checkAndAnd : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        // args[0] is recv
+        // args[1] is the && tmp var
+        // args[2] is the method name
+        // args[3...] are the args
+
+        ENFORCE(args.args.size() >= 3, "Desugar invariant failed");
+        if (args.args.size() < 3) {
+            return;
+        }
+
+        auto selfTy = *args.args[0];
+        auto selfTyAndAnd = *args.args[1];
+
+        if (!isa_type<LiteralType>(args.args[2]->type)) {
+            return;
+        }
+        auto lit = cast_type_nonnull<LiteralType>(args.args[2]->type);
+        if (!lit.derivesFrom(gs, Symbols::Symbol())) {
+            return;
+        }
+        auto fun = lit.asName(gs);
+
+        uint16_t numPosArgs = args.numPosArgs - 3;
+
+        InlinedVector<const TypeAndOrigins *, 2> sendArgStore;
+        InlinedVector<LocOffsets, 2> sendArgLocs;
+        for (int i = 3; i < args.args.size(); ++i) {
+            sendArgStore.emplace_back(args.args[i]);
+            sendArgLocs.emplace_back(args.locs.args[i]);
+        }
+        auto recvLoc = args.locs.args[0];
+        CallLocs sendLocs{args.locs.file, args.locs.call, recvLoc, args.locs.fun, sendArgLocs};
+
+        DispatchArgs innerArgs{
+            fun,
+            sendLocs,
+            numPosArgs,
+            sendArgStore,
+            selfTy.type,
+            selfTy,
+            selfTy.type,
+            args.block,
+            args.originForUninitialized,
+            args.isPrivateOk,
+            args.suppressErrors,
+        };
+        auto dispatched = selfTy.type.dispatchCall(gs, innerArgs);
+
+        auto multipleComponents = dispatched.secondary != nullptr;
+        if (multipleComponents) {
```

Comment:
Why can we bail out on `!multipleComponents`? Is it because nilable is a `T.any`?

Reply:
each component in the DispatchResult corresponds to one branch of the `T.any`

The way that dispatchCall works is that it descends down the structure of the `T.any`, calls `dispatchCall` on each component, and then builds back up a `DispatchResult` that records both sides of the dispatch. Check out the code for `OrType::dispatchCall`

the error message doesn't make sense if there is less than one component—if there is only one component, then we should leave any original error message(s) untouched.